Persistent storage

ruan-andy opened the pull request:
The storage part seems to work but listing the messages in the app does not work. Not sure if that is related to the storage.

alanhzhao commented on src/codeu/chat/client/commandline/Chat.java:
```diff
@@ -47,11 +49,21 @@ public Chat(Context context) {
   // the system wants to exit, the function will return false.
   //
-  public boolean handleCommand(String line) {
+  public boolean handleCommand(String line) throws IOException {
```
This looks like an old version of Chat. Please merge commits from the latest master into this branch and resolve any conflicts.

ruan-andy replied:
Oh, so am I free to merge this into masters?

alanhzhao replied:
No.

    git checkout persistent_storage # if necessary
    git fetch
    git merge origin master

Do this first. This is what I mean by "merge commits from the latest master **into this branch.**" After that I'll review your actual changes.

The reason I'm asking you to do this is because it's harder to separate your changes from changes already in master. (Look at "Files changed.")

Weech4theStars replied:
I backmerged the changes to master into this branch 5 days ago, only difference that wasn't accounted for should be Tammy changing the panel command for version

alanhzhao replied:
That's actually the commit that causes the conflict. I'm guessing you didn't do `git merge`?

In any case, you need to `git merge` now. It's not just me saying this; Github is complaining too.

ruan-andy commented:
Ok, I think I have merge from the origin.

alanhzhao commented on src/codeu/chat/ServerMain.java:
```diff
@@ -89,13 +89,17 @@ public static void main(String[] args) {
                         new RemoteRelay(relaySource);
 
-    final Server server = new Server(id, secret, relay);
+    Server server;
```
Can leave the original lines unchanged.

alanhzhao replied:
please address this

ruan-andy commented:
I have fixed all the merge conflicts. I am unsure of which version we should use because we have changed ServerVersion to ServerInfo on our branch. So I just commented out the ServerVersion just in case we want to switch back to that name.

alanhzhao commented on src/codeu/chat/client/commandline/Chat.java:
```diff
@@ -55,6 +56,12 @@ public boolean handleCommand(String line) throws IOException {
       args.add(token);
     }
+    
+    if(args.size() < 0) {
```
Add space after if

alanhzhao commented on src/codeu/chat/client/commandline/Chat.java:
```diff
@@ -124,10 +131,16 @@ public void invoke(List<String> args) {
       }
     });
     
+    //panel.register("info", new Panel.Command() {
```
Uncomment if necessary, delete if not. Can you make sure both features relying on variations of "ServerInfo" still work?

alanhzhao commented on src/codeu/chat/client/core/Context.java:
```diff
@@ -56,6 +57,8 @@ public UserContext create(String name) {
     return users;
   }
   
+  //public ServerVersion getVersion() {
```
Uncomment if necessary, delete if not.

alanhzhao replied:
Please address this. Generally, we don't need to comment out code; we have revision  history to restore from.

alanhzhao commented on src/codeu/chat/client/core/View.java:
```diff
@@ -35,7 +36,7 @@
 //
 // This is the view component of the Model-View-Controller pattern used by the
-// the client to reterive readonly data from the server. All methods are blocking
+// the client to retrieve readonly data from the server. All methods are blocking
```
Nice catch :)

alanhzhao commented on src/codeu/chat/client/core/View.java:
```diff
@@ -64,6 +65,23 @@ public ServerInfo getInfo() {
   }
+  
+  public ServerVersion getVersion() {
+	    try (final Connection connection = source.connect()) {
```
tabs into spaces

alanhzhao commented on src/codeu/chat/server/Controller.java:
```diff
@@ -34,9 +40,18 @@
 
-  public Controller(Uuid serverId, Model model) {
+  public PrintWriter output;
+  public static boolean writeToLog;
```
public members must be final; otherwise, use a private member with a public getter.

alanhzhao commented on src/codeu/chat/server/Controller.java:
```diff
@@ -34,9 +40,18 @@
+  public PrintWriter output;
+  public static boolean writeToLog;
+
+  public Controller(Uuid serverId, Model model) throws IOException {
```
Avoid throwing checked exceptions in constructors or public methods. Either

* handle the exception, or
* if it can only be caused by programmer error, throw an unchecked exception (a subclass of `RuntimeException`)

alanhzhao commented on src/codeu/chat/server/Controller.java:
```diff
@@ -34,9 +40,18 @@
+    try {
+      output = new PrintWriter(new BufferedWriter(new FileWriter(ServerLog.createFilePath(), true) ) );
+      output.flush();
+    }catch (FileNotFoundException e){
```
add space after `}` and before `{`

alanhzhao commented on src/codeu/chat/server/Controller.java:
```diff
@@ -34,9 +40,18 @@
     this.model = model;
     this.uuidGenerator = new RandomUuidGenerator(serverId, System.currentTimeMillis());
+    try {
+      output = new PrintWriter(new BufferedWriter(new FileWriter(ServerLog.createFilePath(), true) ) );
```
remove spaces between `)`

alanhzhao commented on src/codeu/chat/server/Controller.java:
```diff
@@ -95,6 +110,12 @@ public Message newMessage(Uuid id, Uuid author, Uuid conversation, String body,
       foundConversation.lastMessage = message.id;
     }
+    
+    if(writeToLog) {
```
add space after `if`

alanhzhao commented on src/codeu/chat/server/Controller.java:
```diff
@@ -123,6 +144,11 @@ public User newUser(Uuid id, String name, Time creationTime) {
           creationTime);
     }
+    
+    if(writeToLog) {
```
space after `if`

alanhzhao commented on src/codeu/chat/server/Controller.java:
```diff
@@ -139,6 +165,11 @@ public ConversationHeader newConversation(Uuid id, String title, Uuid owner, Tim
       LOG.info("Conversation added: " + id);
     }
+    
+    if(writeToLog) {
```
space after `if`

alanhzhao replied:
Please address this.

alanhzhao commented on src/codeu/chat/server/Server.java:
```diff
@@ -66,14 +73,44 @@
-  public Server(final Uuid id, final Secret secret, final Relay relay) {
+  public Server(final Uuid id, final Secret secret, final Relay relay) throws IOException {
+  private static ServerInfo info;
+  public PrintWriter output;
```
Either make `final` or change to private w/ public getter.

alanhzhao commented on src/codeu/chat/server/Server.java:
```diff
@@ -66,14 +73,44 @@
-  private static ServerInfo info;
 
-  public Server(final Uuid id, final Secret secret, final Relay relay) {
+  public Server(final Uuid id, final Secret secret, final Relay relay) throws IOException {
```
Avoid throwing checked exceptions in constructors or public methods. Either

* handle the exception, or
* if it can only be caused by programmer error, throw an unchecked exception (a subclass of `RuntimeException`)

ruan-andy replied:
Hi Alan, I am confused about handling the exception. I thought when we put the code in the try block and catching the exception is when we are handling it.

alanhzhao replied:
Basically, this constructor should not have `throws IOException`. The two options I suggested above help you resolve it. I would catch the `IOException` and throw a `RuntimeException` because there is no real way to recover.

(I also think the over-use of `IOException` in this code base is an anti-pattern, but oh well.)

alanhzhao replied:
Please address this. If you need more clarification, ping me on slack.

alanhzhao commented on src/codeu/chat/common/ServerVersion.java:
```diff
@@ -0,0 +1,28 @@
+
+import codeu.chat.util.Uuid;
+
+public final class ServerVersion {
```
Describe class with javadoc.

alanhzhao commented on src/codeu/chat/server/Server.java:
```diff
@@ -66,14 +73,44 @@
+    info = new ServerInfo();
+    version = new ServerVersion();
+
+//    try {
```
Uncomment if necessary; remove if not.

alanhzhao commented on src/codeu/chat/server/Server.java:
```diff
@@ -292,4 +329,11 @@ public void run() {
 
+  // checks if the server needs restoring
+  private void restore(ServerLog log, Controller controller) {
+	for(int i = 0; i < log.getLength(); i++) {
```
add space after `for`

alanhzhao commented on src/codeu/chat/util/ServerLog.java:
```diff
@@ -0,0 +1,120 @@
+
+public final class ServerLog {
+  private File log;	
+  private HashMap<Integer,String> lines;
```
Use `Map` instead of `HashMap` here. Generally, it's better to use interface classes for variables. Example:

`List<Integer> = new ArrayList<>();`

alanhzhao commented on src/codeu/chat/util/ServerLog.java:
```diff
@@ -0,0 +1,120 @@
+	  return Time.parse(s);
+  }
+
+  private HashMap<Integer,String> readFile(BufferedReader br) {
```
`Map`

alanhzhao commented on src/codeu/chat/util/ServerLog.java:
```diff
@@ -0,0 +1,120 @@
+
+  private HashMap<Integer,String> readFile(BufferedReader br) {
+    String line;
+    HashMap<Integer,String> lines = new HashMap<Integer,String>();
```
`Map<...> = new HashMap<...>()`

alanhzhao commented on src/codeu/chat/util/ServerLog.java:
```diff
@@ -0,0 +1,120 @@
+    char commandType = toParse.charAt(0);
+
+    try {
+      if(commandType == 'M') {
```
add space after `if`

alanhzhao commented on src/codeu/chat/util/ServerLog.java:
```diff
@@ -0,0 +1,120 @@
+
+    // turns the string from log into an array
+    String[] ParArr = toParse.split("_");
+    char commandType = toParse.charAt(0);
```
what if `toParse` is empty?

alanhzhao commented on src/codeu/chat/util/ServerLog.java:
```diff
@@ -0,0 +1,120 @@
+        controller.newConversation(Uuid.parse(ParArr[2]), ParArr[1], 
+		                 Uuid.parse(ParArr[3]), stringToTime(ParArr[4]) );
+        }
+    } catch(IOException e) {
```
space after `catch`

alanhzhao commented on test/codeu/chat/server/BasicControllerTest.java:
```diff
@@ -32,7 +35,11 @@
+    try {
+		controller = new Controller(Uuid.NULL, model);
+	} catch (IOException e) {
+		e.printStackTrace();
```
throw an unchecked exception so the test fails

alanhzhao commented on test/codeu/chat/relay/logTest.java:
```diff
@@ -0,0 +1,75 @@
+	try {
+	  controller = new Controller(Uuid.NULL, model);
+	} catch (IOException e) {
+			e.printStackTrace();
```
throw an unchecked exception so the test fails

ruan-andy commented:
Hey Alan, I have fixed all the requested changes but not sure if I fixed the requested uncheck exception changes. Can you look over that. Thanks

alanhzhao commented on src/codeu/chat/client/commandline/Chat.java:
```diff
@@ -55,6 +56,12 @@ public boolean handleCommand(String line) throws IOException {
       args.add(token);
     }
+    
+    if (args.size() < 0) {
```
This condition is always false. You want `size == 0`.

alanhzhao commented on test/codeu/chat/relay/logTest.java:
```diff
@@ -0,0 +1,69 @@
+    assertTrue(relay.addTeam(team, secret));
+
+    assertTrue(relay.write(team, 
+               	           secret,
```
Tabs > spaces

alanhzhao commented on test/codeu/chat/server/RawControllerTest.java:
```diff
@@ -37,7 +40,11 @@
     model = new Model();
-    controller = new Controller(Uuid.NULL, model);
+    try {
+		controller = new Controller(Uuid.NULL, model);
```
tabs > spaces

alanhzhao commented on src/codeu/chat/util/ServerLog.java:
```diff
@@ -0,0 +1,124 @@
+      if (commandType == 'M') {
+        // parse a message
+        controller.newMessage(Uuid.parse(ParArr[2]), Uuid.parse(ParArr[1]), 
+		            Uuid.parse(ParArr[3]), ParArr[5], stringToTime(ParArr[4]) );
```
Fix indentation: remove tabs, add spaces.

alanhzhao commented on src/codeu/chat/server/Server.java:
```diff
@@ -292,4 +322,11 @@ public void run() {
 
+  // checks if the server needs restoring
+  private void restore(ServerLog log, Controller controller) {
+	for (int i = 0; i < log.getLength(); i++) {
```
Fix indentation: remove tabs, add spaces.

alanhzhao commented on src/codeu/chat/ServerMain.java:
```diff
@@ -89,13 +89,17 @@ public static void main(String[] args) {
                                 Secret secret,
                                 ConnectionSource serverSource,
-                                ConnectionSource relaySource) {
+                                ConnectionSource relaySource) throws IOException {
```
After you change the `Server` constructor, remove `throws IOException` here.

alanhzhao commented:
Since #7 was merged just now, you'll need to merge from master again.

adsnaider commented on src/codeu/chat/client/commandline/Chat.java:
```diff
@@ -128,6 +135,7 @@ public void invoke(List<String> args) {
       @Override
       public void invoke(Scanner args) {
```
I don't think this compiles

adsnaider commented on src/codeu/chat/common/ServerVersion.java:
```diff
@@ -0,0 +1,31 @@
+package codeu.chat.common;
```
I think all of this file is pretty much alongside ServerInfo.java
No need to split it up

adsnaider commented on src/codeu/chat/server/Server.java:
```diff
@@ -65,15 +72,38 @@
+    
+    codeu.chat.server.Controller.writeToLog = false;
+
+    info = new ServerInfo();
```
No need. It's defined up above :)

adsnaider commented on src/codeu/chat/server/Server.java:
```diff
@@ -65,15 +72,38 @@
+      }
+    });
 
     info = new ServerInfo();
```
This should probably not be here
